### ISSUE TITLE
Dimensional patch deserialization performanc issue

### ISSUE THREAD

**Issue by strk** [pgpointcloud/pointcloud#42]
Running some performance tests I was surprised to find deserialization of dimensionally-compressed patches being as slow as decompression of PostgreSQL compressed datums (main storage).

Reading the code I found a useless call to pc_bytes_serialized_size in pc_patch_dimensional_deserialize.

Removing the useless call I get a performance improvement of ~60% (deserialization runs in ~1/10 of the time) for a patch with 6.5 millions of points and 8 dimensions of 8 bytes each.

**Comment by Remi-C** [comment]
Hm that would explain a lot (like the fact that using compressed or not
TOAST doesn't change export time).
It could aslo explain why decompression is "slow " (even with good cpu, my
output is bound to 100k pts/sec)

Is it possible to test this yet?

Cheers,
Rémi-C

2015-02-13 13:43 GMT+01:00 Sandro Santilli <notifications@github.com>:

**Comment by strk** [comment]
Sure, go ahead! It's in master and in 1.0.x branches

**Comment by strk** [comment]
from 11811.655 to 3309.709 ms for me to Filter 65341 patches of 1266 points each stored with dimensional compression (one of many cases)

**Comment by Remi-C** [comment]
Very promising indeed.
I'll try it this week end.
Tanks
Rémi-C

2015-02-13 17:05 GMT+01:00 Sandro Santilli <notifications@github.com>:

**Comment by strk** [comment]
Re-opening, must be fixed in another way (see #44)

**Comment by strk** [comment]
on a closer look, the pc_bytes_serialized_size() function implementation is really simple, so now I wonder if the the performance improvement I was observing was really due to the bare function call cost or to something about further compression in the pipeline which made the dataset smaller (due to more equal values)

**Comment by strk** [comment]
I've tried fully inlining the function call but that does not improve deserialization speed at all.
And I confirm that avoiding the moving of buf pointer results in timings which are on the order of 3 seconds instead of 16 seconds. Impressive, but I guess it has to do with the kind of compression engaged by the different dimensions, with the first dimension (always hit by the broken code) being the fastest.

It'd take a function to inspect the kind of compression for each dimension to know more about this case.

**Comment by strk** [comment]
All but last dimension have compression PC_DIM_SIGBITS, last one has PC_DIM_ZLIB. 
So new guess is that ZLIB compression is 5 times slower than SIGBITS.
```
NOTICE:  Compression for dim 0: 2
NOTICE:  Compression for dim 1: 2
NOTICE:  Compression for dim 2: 2
NOTICE:  Compression for dim 3: 3
```

**Comment by Remi-C** [comment]
ZLib being slower would be a reasonnable hypothesis isn't it.

So as a side effect we can infer that use float instead of int has a cost
in term of memory space and compression time.


Cheers,
Rémi-C

2015-03-10 17:03 GMT+01:00 Sandro Santilli <notifications@github.com>:

**Comment by strk** [comment]
Yep, reasonable hypothesis. Which compression is used when can be seen here:
https://github.com/pgpointcloud/pointcloud/blob/v0.1.0/lib/pc_dimstats.c#L109-L138

PC_DIM_RLE compression was not tested in this case.

**Comment by strk** [comment]
Actually, in this case dimension 4 is an ``uint16_t`` interpretation. 2 bytes integer.
So evidently raw/sigbits <= 1.6 and raw/rle <= 4.0. Now I wonder why 2 different values are used there, and what's the actual compression factor for zlib.

**Comment by strk** [comment]
is there a function to drop and re-compute stats ?

**Comment by Remi-C** [comment]
It seems you can free stats then recompute it

    pc_dimstats_free
    pc_patch_uncompressed_compute_stats

You might have to uncompress the patch before computing new stats.

Rémi-C

2015-03-10 17:45 GMT+01:00 Sandro Santilli <notifications@github.com>:

**Comment by strk** [comment]
I meant from the SQL level.

The pc_uncompress function does not change the patch "type" so serialization skips compression as it thinks the patch is already compressed (!). Is this intended behavior @pramsey ?

**Comment by strk** [comment]
Found the SQL way: pc_patch(pc_explode())

**Comment by strk** [comment]
So here are the dimensional compressions stats for my ~1275 points patches:
```
NOTICE:  dim:0 -- raw/sigbits:4.53888 -- raw/rle:1.0119
NOTICE:  dim:1 -- raw/sigbits:3.53583 -- raw/rle:1.01291
NOTICE:  dim:2 -- raw/sigbits:2.27755 -- raw/rle:0.8
NOTICE:  dim:3 -- raw/sigbits:0.998434 -- raw/rle:1.08418
```

**Comment by strk** [comment]
For completeness, the zlib compression factor for dim:3 -- NOTICE:  raw/zlib:6.21951

**Comment by strk** [comment]
Alright, closing this as "expected".

**Comment by Remi-C** [comment]
So what is the end of it finally?
Improvement, or no improvement possible because it introduces a bug?

**Comment by strk** [comment]
Cannot exclude improvements are possible, but the idea of skipping
the call to pc_bytes_serialized_size was not a good one, as the
function is really short and easy...

**Comment by strk** [comment]
I'm reopening this with latest findings about speed issues.
Now that I have PC_Compress() function I could check the time it takes to _compress_ with different methods, and zlib again turned out to be very very expensive.

To compress 8232 patches uncompressed patches each with ~1000 points of 12 dimensions, we have:
  - zlib  29 seconds
  - rle    2 seconds
  - sigbits   1.5 seconds

**Comment by strk** [comment]
Maybe we can make the ZLIB compression level configurable. Right now it's hard-coded to max (9)

**Comment by pramsey** [comment]
Why? The CPU difference is teeny, and smaller is better? 
Signed: the Alliance for Fewer Knobs.

**Comment by strk** [comment]
I run a new test. ZLIB-compressing each of 9 dimensions in 10000 patches of ~600 points each (real world data, from http://dati.toscana.it/dataset/lidar/resource/b057f578-cb63-49bc-b050-75fcb808d1d2).

Compression level 9 completes in 12.4 seconds and results in 64,479,226 bytes total.
Compression level 1 completes in 12.0 seconds and results in 67,656,275 bytes total.

I guess it gets more interesting as the number of points in the patches grow.

For the record, compressing the same patches with the automatically determined dimensional compression completes in 3.7 seconds and results in 63,522,825 bytes total.

**Comment by strk** [comment]
Another bit of information, the "automatically determined dimensional compression" still uses ZLIB for one of the dimensions. It indeed takes 3.7 seconds with compression level 1 and 4.9 seconds with compression level 9. For comparison, compressing all dimensions with "sigbits" takes 1.6 seconds.
I wonder if "dimensional compression" should allow for uncompressed dimensions sometimes.

I think PostgreSQL does check if compression is worth it against an initial relatively small portion of the data, before fully compressing it.
